### PR TITLE
Header styling & paper-tabs improvements

### LINKED ISSUE #6985: Jumping panel title + empty space
## Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

## The problem

When switching to another panel, the title is jumping to the right since the hidden button space gets added. 

Can't we get rid of the hidden empty icon space (`<ha-menu-button>`), if a separate side bar is shown, since then the drawer / burger icon is hidden anyway but still takes up space (which looks wrong independent of the jumping)? So perhaps some dynamic overlay CSS magic that if the screen is wide enough to display the sidebar, then `ha-menu-button` allows the title element to overlay it?

B6F29966-5BDA-4292-B73E-F612C0173B32

## Expected behavior

1. No jumping title
2. No empty space in front of the title (or perhaps show the panel icon from the sidebar again, but I did not see an easy way to get access to the icon code other than hard-coding / duplicating it in the panel code itself which feels wrong)

## Environment

- Home Assistant release with the issue: latest dev (0.115)
- Last working Home Assistant release (if known):
- Browser and browser version: FF and Chrome
- Operating system: Win 10

## State of relevant entities

```yaml

```

## Problem-relevant configuration

```yaml

```

## Javascript errors shown in your browser console/inspector

```txt

```

### LINKED ISSUE #7107: Bring visual alignment between developer and configuration menus
## Bring visual alignment between developer and configuration menus 

It would be great to have a similar look and feel between the Developer and the Configuration Menus. It currently looks like these two were designed by completely different teams (which could be entirely possible).

While being cognizant that the overall design should not make too many clicks to get to the services tab, but this would only add one click to get to the services screen

**TODAY:**
image

**POSSIBLE:**
image

## Additional information

I have the Chrome Inspect window open but can’t copy the HTML code because of what I think are these shadow-root errors.  If anyone knows how to fix this, LMK.

### PR DESCRIPTION
**Pull request by maykar**
## Proposed change

This changes the styling of the header across most of the frontend. It removes the large amount of space on the left side of the screen and allows tabs to span the width of the header.

Additionally, this adds an ha-tabs element to modify paper-tabs. It adds:

* Optimized space for scroll arrows (keeps them "inside" the scrolling container element and only take up space when needed) 
* Tweaks to scrolling to make it fluid (no jumping while scrolling when scroll arrows hide/show).
* Scrolls active tab into view if needed.

This is also to help pave the way for compacting the header later on.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

drawing

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/components/ha-menu-button.ts [hunk 1]
```diff
@@ -130,6 +130,7 @@ class HaMenuButton extends LitElement {
     return css`
       :host {
         position: relative;
+        margin-right: 13px;
```
Comment (bramkragten) [inline src/components/ha-menu-button.ts:133]: We should stick to material guidelines, that means all padding/margin is a multiple of 4px.

So make this 12px

**Comment by bramkragten** [comment]
The `main-title` needs margin left, I think 16px.

File: src/panels/lovelace/hui-root.ts [hunk 2]
```diff
@@ -699,7 +700,7 @@ class HUIRoot extends LitElement {
           min-height: 100%;
         }
         paper-tabs {
-          margin-left: max(env(safe-area-inset-left), 12px);
+          margin-left: max(env(safe-area-inset-left), 17px);
           margin-right: env(safe-area-inset-right);
```
Comment (bramkragten) [inline src/panels/lovelace/hui-root.ts:704]: Left and right should be equal? (and multiple of 4)

**Comment by maykar** [comment]
For main title I only removed `app-toolbar ha-menu-button + [main-title]`

This remains
```
  app-toolbar ha-icon-button-arrow-prev + [main-title],
  app-toolbar ha-icon-button + [main-title] {
    margin-left: 24px;
  }
```

**Comment by maykar** [comment]
The `main-title` margin for those might need adjusted to 12px as well, though. To be consistent. Will double check how that looks.

**Comment by bramkragten** [comment]
> For `main-title` I only removed `app-toolbar ha-menu-button + [main-title]`
> 
> This remains
> 
> ```
>   app-toolbar ha-icon-button-arrow-prev + [main-title],
>   app-toolbar ha-icon-button + [main-title] {
>     margin-left: 24px;
>   }
> ```

Yes, but the `[main-title]` was never shown with an element in front. The title should align with the tabs.

File: src/components/ha-menu-button.ts [hunk 3]
```diff
@@ -130,6 +130,7 @@ class HaMenuButton extends LitElement {
     return css`
       :host {
         position: relative;
+        margin-right: 12px;
```
Comment (bramkragten) [inline src/components/ha-menu-button.ts:133]: This should not be in `ha-menu-button` the user of this element should set the margin.

Comment (bramkragten) [inline src/components/ha-menu-button.ts:133]:
```suggestion
```

File: src/resources/styles.ts [hunk 4]
```diff
@@ -110,10 +110,9 @@ export const haStyle = css`
     color: var(--app-header-text-color, white);
   }
 
-  app-toolbar ha-menu-button + [main-title],
   app-toolbar ha-icon-button-arrow-prev + [main-title],
   app-toolbar ha-icon-button + [main-title] {
-    margin-left: 24px;
+    margin-left: 12px;
   }
```
Comment (bramkragten) [inline src/resources/styles.ts:116]:
```ts
  app-toolbar [main-title] {
    margin-left: 12px;
  }
```

**Comment by maykar** [comment]
Updated gif above with up to date changes

File: src/panels/developer-tools/ha-panel-developer-tools.ts [hunk 5]
```diff
@@ -107,7 +108,7 @@ class PanelDeveloperTools extends LitElement {
           height: calc(100vh - 112px);
         }
         paper-tabs {
-          margin-left: 12px;
+          margin-left: 24px;
```
Comment (bramkragten) [inline src/panels/developer-tools/ha-panel-developer-tools.ts:111]:
```suggestion
          margin-left: max(env(safe-area-inset-left), 24px);
          margin-right: max(env(safe-area-inset-right), 24px);
```

**Review by bramkragten — APPROVED** [review]
Looks good!

**Comment by bramkragten** [comment]
Checking the material guidelines, I wonder we can could offset the items, or show an icon to show that there are more items

https://material.io/components/tabs#scrollable-tabs

**Comment by maykar** [comment]
Not so sure about offsetting, but indicator would be nice. Scroll listener and hide/show indicators if paper-tabs can still be scrolled left/right. Will mess around with this.

**Comment by maykar** [comment]
Unfortunately, I don't think custom indicators are possible without some pretty hacky script after render. `paper-tabs` has both the indicator elements and the tabContainer element that does the actual scrolling are in the shadowDOM with no accessors.

**Comment by maykar** [comment]
The mention of offsetting the items in the material guidelines is either trying to illustrate that the last item would be cut off to indicate scrolling (which we couldn't guarantee) or just that the offset itself suggests it.

The offset was like that before when I was lining up the header text with the start of the icon/beginning of the first tab indicator line, but we changed it to make text within the first tab or start of the icon line up with header.

I'm a bit torn on the issue, while it would be nice to have an indication of scrolling it's not really possible with the state of `paper-tabs`. Customizing the paper-tabs element for our purposes isn't out of the question, but if this would be replaced with MWC tabs later on it may not be worth it.

**Comment by maykar** [comment]
After a fair bit of testing I was able to get working indicators, but only by calculating the bounds of the `paper-tabs` and the first & last `paper-tab` elements. The only way to get `paper-tabs` to do this was by doing the calculations on every mouse move (`onmousemove`). So, as far as the `paper-tabs` element, I think we're out of luck as far as scroll indicators are concerned. Will look into MWC tabs and see if that is a more viable option.

**Comment by maykar** [comment]
Added a modified version of paper-tabs that will work with these changes while still having indicators/scroll buttons. You can see the modifications made here: https://github.com/PolymerElements/paper-tabs/compare/master...maykar:master

File: src/components/paper-tabs/paper-tabs.js [hunk 6]
```diff
@@ -0,0 +1,663 @@
+/**
```
Comment (bramkragten) [inline src/components/paper-tabs/paper-tabs.js:1]: Don't add a copy of `paper-tabs`, but create a `ha-paper-tabs` that extends `paper-tabs` and adjust the things that need to be adjusted.

File: src/components/ha-tabs.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      } else {
+        return;
+      }
+      e.preventDefault();
+    });
+  }
+
+  _affectScroll(dx) {
+    super._affectScroll(dx);
+    this.$.tabsContainer.scrollLeft += dx;
```
Comment (bramkragten) [inline src/components/ha-tabs.js:42]: We do this twice now, is that intentional?

File: src/components/ha-tabs.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    const scrollLeft = this.$.tabsContainer.scrollLeft;
+
+    this._leftHidden = scrollLeft - this.firstElementChild.clientWidth < 0;
```
Comment (bramkragten) [inline src/components/ha-tabs.js:46]: `this.firstElementChild.clientWidth` is expensive, do we need to do that on every scroll?

File: src/components/ha-tabs.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      } else if (e.wheelDelta < 0) {
+        this._scrollToLeft();
+      } else {
+        return;
+      }
+      e.preventDefault();
+    });
+  }
+
+  _affectScroll(dx) {
```
Comment (bramkragten) [inline src/components/ha-tabs.js:40]: Can you explain in a comment why we adjust this function?

Comment (maykar) [reply src/components/ha-tabs.js:40]: When I adjusted the scroll tabs to take up no more space than the actual tab container, it meant that the tab container actually shrinks when the scroll arrows appear. When that happens the area that you were scrolling jumps by the width of the scroll arrow. This prevents that from happening by jumping in the opposite direction the same amount, giving the appearance that nothing has changed and the scroll is smooth.

File: src/components/ha-tabs.js (hunk 8 above)
Comment (maykar) [reply src/components/ha-tabs.js:46]: No, we could calculate this on ready

File: src/components/ha-tabs.js (hunk 7 above)
Comment (maykar) [reply src/components/ha-tabs.js:42]: Not intentional

File: src/components/ha-tabs.js (hunk 8 above)
Comment (maykar) [reply src/components/ha-tabs.js]: calculating on _tabsChanged now

File: src/components/ha-tabs.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    this.addEventListener("mousewheel", (e) => {
+      if (e.wheelDelta > 0) {
+        this._scrollToRight();
+      } else if (e.wheelDelta < 0) {
+        this._scrollToLeft();
+      } else {
+        return;
+      }
+      e.preventDefault();
+    });
```
Comment (bramkragten) [inline src/components/ha-tabs.js:36]: Let's remove the scrolling (at least for now)

**Comment by maykar** [comment]
Updated gif in PR description with current changes.

File: src/components/ha-tabs.js [hunk 11]
```diff
@@ -10,6 +10,12 @@ class HaTabs extends PaperTabsClass {
 
       const superStyle = subTemplate.content.querySelector("style");
 
+      [...subTemplate.content.querySelectorAll("paper-icon-button")].forEach(
+        (arrow) => {
+          arrow.setAttribute("noink", "");
```
[the remark this answers is not in the document]
Comment (maykar) [reply src/components/ha-tabs.js:15]: I'm not removing the attributes value, I'm creating an empty attribute called noink. This tells the element to not use the ripple animation.

[the remark this answers is not in the document]
Comment (maykar) [reply src/components/ha-tabs.js:16]: Added comment to clarify.

File: src/components/ha-tabs.ts [hunk 12]
```diff
@@ -36,10 +36,11 @@ class HaTabs extends PaperTabsClass {
...
   _tabChanged(tab, old) {
     super._tabChanged(tab, old);
-    
+
     // Get first and last tab's width for _affectScroll
-    this.firstTabWidth = this.firstElementChild.clientWidth;
-    this.lastTabWidth = this.lastElementChild.clientWidth;
+    const tabs = this.querySelectorAll("paper-tab:not(.hide-tab)");
+    this.firstTabWidth = tabs[1].clientWidth;
+    this.lastTabWidth = tabs[tabs.length - 1].clientWidth;
```
Comment (bramkragten) [inline src/components/ha-tabs.ts:43]: What if I only have 1 tab?

File: src/components/ha-tabs.ts [hunk 13]
```diff
@@ -36,10 +36,11 @@ class HaTabs extends PaperTabsClass {
 
   _tabChanged(tab, old) {
     super._tabChanged(tab, old);
-    
+
     // Get first and last tab's width for _affectScroll
-    this.firstTabWidth = this.firstElementChild.clientWidth;
-    this.lastTabWidth = this.lastElementChild.clientWidth;
+    const tabs = this.querySelectorAll("paper-tab:not(.hide-tab)");
+    this.firstTabWidth = tabs[1].clientWidth;
```
Comment (bramkragten) [inline src/components/ha-tabs.ts:42]: Should be the first right?
```suggestion
    this.firstTabWidth = tabs[0].clientWidth;
```

File: src/panels/lovelace/editor/view-editor/hui-dialog-edit-view.ts [hunk 14]
```diff
@@ -158,10 +158,8 @@ export class HuiDialogEditView extends LitElement {
         .heading=${true}
       >
         <div slot="heading">
-          <h2>
-            ${this._viewConfigTitle}
-          </h2>
-          <paper-tabs
+          <h2>${this._viewConfigTitle}</h2>
+          <ha-tabs
```
Comment (bramkragten) [inline src/panels/lovelace/editor/view-editor/hui-dialog-edit-view.ts:162]: This should use `mwc-tabs` and not `ha-tabs`

File: src/components/ha-tabs.ts [hunk 15]
```diff
@@ -1,16 +1,28 @@
...
+
   static get template() {
     if (!subTemplate) {
-      subTemplate = PaperTabsClass.template.cloneNode(true);
+      subTemplate = PaperTabs.template.cloneNode(true);
 
       const superStyle = subTemplate.content.querySelector("style");
 
-      // Add "noink" attribute to scroll buttons to disable click animation.
       [...subTemplate.content.querySelectorAll("paper-icon-button")].forEach(
```
Comment (bramkragten) [inline src/components/ha-tabs.ts:26]: Why make a copy of the array?

File: src/components/ha-tabs.ts [hunk 16]
```diff
@@ -54,7 +67,11 @@ class HaTabs extends PaperTabsClass {
    * while scrolling and the tab container shrinks we can counteract
    * the jump in tab position so that the scroll still appears smooth.
    */
-  _affectScroll(dx) {
+  public _affectScroll(dx) {
```
Comment (bramkragten) [inline src/components/ha-tabs.ts:70]:
```suggestion
  public _affectScroll(dx: number):void {
```

File: src/components/ha-tabs.ts [hunk 17]
```diff
@@ -1,16 +1,28 @@
...
 let subTemplate;
 
-class HaTabs extends PaperTabsClass {
+@customElement("ha-tabs")
+export class HaTabs extends PaperTabs {
+  @property({ type: Number }) public firstTabWidth = 0;
+
+  @property({ type: Number }) public lastTabWidth = 0;
+
+  @property({ type: Boolean }) public _lastLeftHiddenState = false;
```
Comment (bramkragten) [inline src/components/ha-tabs.ts:18]: The should be `private` and `internalProperty` also prepend private properties with an `_`

Comment (bramkragten) [reply src/components/ha-tabs.ts:18]: But... you can't use properties in Polymer elements, this is for LitElements.

Comment (bramkragten) [inline src/components/ha-tabs.ts:18]:
```suggestion
  private _firstTabWidth = 0;

  private _lastTabWidth = 0;

  private _lastLeftHiddenState = false;
```

File: src/components/ha-tabs.ts [hunk 18]
```diff
@@ -34,13 +46,14 @@ class HaTabs extends PaperTabsClass {
...
+  // Get first and last tab's width for _affectScroll
+  public _tabChanged(tab, old) {
     super._tabChanged(tab, old);
-
-    // Get first and last tab's width for _affectScroll
     const tabs = this.querySelectorAll("paper-tab:not(.hide-tab)");
-    this.firstTabWidth = tabs[0].clientWidth;
-    this.lastTabWidth = tabs[tabs.length - 1].clientWidth;
+    if (tabs.length > 1) {
+      this.firstTabWidth = tabs[1].clientWidth;
```
Comment (bramkragten) [inline src/components/ha-tabs.ts:54]:
```suggestion
      this.firstTabWidth = tabs[0].clientWidth;
```

File: src/components/ha-tabs.ts [hunk 19]
```diff
@@ -34,13 +46,14 @@ class HaTabs extends PaperTabsClass {
...
-  _tabChanged(tab, old) {
+  // Get first and last tab's width for _affectScroll
+  public _tabChanged(tab, old) {
     super._tabChanged(tab, old);
-
-    // Get first and last tab's width for _affectScroll
     const tabs = this.querySelectorAll("paper-tab:not(.hide-tab)");
-    this.firstTabWidth = tabs[0].clientWidth;
-    this.lastTabWidth = tabs[tabs.length - 1].clientWidth;
+    if (tabs.length > 1) {
```
Comment (bramkragten) [inline src/components/ha-tabs.ts:53]:
```suggestion
    if (tabs.length) {
```

File: src/components/ha-tabs.ts [hunk 20]
```diff
@@ -0,0 +1,93 @@
+import { customElement } from "lit-element";
+import { PaperTabsElement } from "@polymer/paper-tabs/paper-tabs";
```
Comment (bramkragten) [inline src/components/ha-tabs.ts:2]:
```suggestion
import "@polymer/paper-tabs/paper-tabs";
import type { PaperTabsElement } from "@polymer/paper-tabs/paper-tabs";
```

File: src/panels/lovelace/editor/view-editor/hui-dialog-edit-view.ts (hunk 14 above)
Comment (maykar) [reply src/panels/lovelace/editor/view-editor/hui-dialog-edit-view.ts:162]: Do you mean keep as paper-tabs?

Comment (maykar) [reply src/panels/lovelace/editor/view-editor/hui-dialog-edit-view.ts:162]: I stand corrected

**Review by bramkragten — APPROVED** [review]

**Comment by poldim** [comment]
@maykar are you sure you meant fixes #7107? That wasn’t about the header but rather page layout. 

BTW, I’m hoping hiding the header is made into the core since CCH is no longer supported!

**Comment by maykar** [comment]
Commented on the original issue, it was a misunderstanding on my part, tagged bram to reopen.

**Pull request merged**